kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,215 @@
+
+On the other hand, from the cluster administrative view, unconditional non-preempted pods are not desirable because there are system-critical, node-critical pods in the cluster. So, the cluster administrator would like to set the minimum priority value for such a priority class.  For example, assume there are three priority classes: system-critical, high and low.  Then, the cluster administrator would like to introduce `low-non-preempted` priority class that can not be preempted by high but can be preempted by system-critical.
+
+```yaml
+  system-critical: 10000
+             high:  9000
+              low:  8000
+low-non-preempted:  8000 # with minimumPreemptablePriority=10000
+                        # (i.e. high can't preempt this, but system-critical can preempt this)
+```
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. As described above, job would need check-pointing. However, most machine learning jobs are iteration-based process, called _epochs_. So, if the cluster is high load and preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and computational resources consumed by the job came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+type PreemptionToleration struct {
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  // If not set, pods that have a higher priority value can preempt it (default).
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.
+  // Nil value means forever, which is not preempted forever (default)
+  // zero and negative value means immediate, which is preempted immediately. i.e. no toleration at all.
```

Comment:
I don't think we should distinguish nil from zero. How about:

```
// It defaults to zero if not set. Zero value means the pod will be preempted immediately. i.e., no toleration at all.
```

And negative value will be caught during validation phase.

Reply:
As stated https://github.com/kubernetes-sigs/scheduler-plugins/pull/205#discussion_r662106942:

- `TolerationSeconds: nil` : it can tolerate forever (_perhaps this is a common use case, I think_)
- `TolerationSeconds: 0`: it can't tolerate at all (tolerate in 0 second).

So, how about this??

```
  // TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  
  // It defaults to Nil if not set.  Nil value means the pod can tolerate the preemption forever.  Zero value means the pod will be preempted immediately. i.e., no toleration at all.
```

Reply:
updated in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/671bf6ff7c31eac20f83555497028ccac4852f7a

Reply:
@everpeace in terms of API design practices, it's a good practice to use a pointer for the internal-version struct (in pkg/apis/config/types.go); we do that for external versions (in pkg/apis/config/v1XYZ/types.go).

So when the versioned obj is wired in, we can tell if it's set to 0 or left as nil, and then default the value if it's nil (not set). In the internal processing that uses internal version objs, we don't need to worry the fields are set or not - default or user-specified value should have been set.

So the comment "It defaults to Nil if not set" is inaccurate - in internal versions, it should be a concrete value instead of pointer.

Reply:
@Huang-Wei Ah, I understood.  Then, how about negative value means forever??  I would like to implement a policy that can tolerate _forever_ in some way. How about below??

```
// TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  
// It defaults to zero if not set. Zero value means the pod will be preempted immediately. i.e., no toleration at all.
// Negative value means the pod can tolerate forever.
```

done in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/b15a95027b612c9b8c8682d2ce2833bc03fd7f81

Reply:
OR, could we make `Nill` be default (perhaps we can say no default). how about this??

```
// TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  
// If not set, the pod can tolerate the preemption forever.
```

Reply:
> just leave`Nil` if not set

As I mentioned before, internal API object should be a concrect value (so nil value doesn't work, generally).

You previous comment makes sense to me:

- if it's not set (nil), internally default it to 0 - which means it can be preempted immediately
- if it's set to a positive value - the duration will be honored
- if it's set to a negative value, you _can_ make it semantically mean "don't preempt forever". But I'm not sure this is really what you want? What if some pods set this to a negative value and then the high-priority can never preempt them and pending there forever? It seems to break the semantics of Priority.

Reply:
@Huang-Wei 
> As I mentioned before, internal API object should be a concrect value (so nil value doesn't work, generally).

oh... sorry... I was confused.

>  But I'm not sure this is really what you want? What if some pods set this to a negative value and then the high-priority can never preempt them and pending there forever? It seems to break the semantics of Priority.

`minimumPreemptablePriority` and `tolerationSeconds` will be evaluated in AND-ed.  `tolerationSeconds` is expected to be used with `minimumPreemptablePriority`.   It means `tolerationSeconds` specifies how long this priority class can tolerate preemption by priorities __lower than MinimumPreemptablePriority__ as described in the comment.  For example,

Assume `low-non-preempted-10min` declared with `minimumPreemptablePriority=10000, tolerationSeconds=600`.  Then, `system-critical, high, medium` priorites' preemption behavior will be like below:

```yaml
# system-critical can preempt low-10min pods immediately 
# because low-non-preempted-10min can't tolerate the preemption from this priority class
# (because of minimumPreemptablePriority=10000)
         system-critical: 10000 

# high and medium pods can preempt low-non-preempted-10min pods 
# which elapsed at least 10min from being scheduled because toleration expired in 10 min
# (because of tolerationSeconds=600)
                   high:  9000 
                 medium:  8500 

# assume low-non-preempted-10min has a preemption toleration policy 
# with minimumPreemptablePriority=10000, tolerationSeconds=600
low-non-preempted-10min:  8000
```

updated in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/5ab5cad9eccc3da671bbfecdffa874ee14d1e866

Reply:
I thought `minimumPreemptablePriority` and `tolerationSeconds` work independently. The current semantics looks good - it can only tolerate the pods with priority lower than  `minimumPreemptablePriority`, not always. Can you somehow emphasize it in the example? You did mention it in the latter section (API/implementations), but it'd be better to also emphasize it in the earlier section (the examples).

Reply:
@Huang-Wei 
> Can you somehow emphasize it in the example? You did mention it in the latter section (API/implementations), but it'd be better to also emphasize it in the earlier section (the examples).

Thanks for your suggestion. I agree.  I think it would the earlier "Use Cases" section would be better not to require the readers preemption toleration API knowledge.  So, I added the section "Implementing Typical Use Cases by Preemption Toleration APIs" under the "Preemption Toleration API" section so that it can describe how the API works in detail by examples. And I added the link to the section in "Use Cases" section. 

updated in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/c50b6e20037a141f1cd9eda22fd7ed2aeaed926f .